Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76477
* #76360

Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

pritamdamania87 committed:
Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor.

Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor."


Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor."


Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

fduwjj commented on test/distributed/_shard/sharded_tensor/ops/test_softmax.py:
```diff
@@ -0,0 +1,57 @@
+        st = _shard_tensor(local_tensor, spec)
+        sharded_softmax = torch.nn.functional.softmax(st, dim)
+
+        self.assertEqual(local_softmax.chunk(self.world_size)[self.rank], sharded_softmax.local_tensor())
```
Nit: you can just use _shard_tensor and directly compare two sharded tensor.

pritamdamania87 replied:
`.chunk` seems simpler for a test rather than sharding the local tensor :)

fduwjj commented on torch/distributed/_shard/sharded_tensor/partial_tensor.py:
```diff
@@ -152,3 +184,61 @@ def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
+            f"torch function '{func.__name__}', with args: {args} and "
+            f"kwargs: {kwargs} not supported for PartialTensor!")
+
+    def transpose(self, dim0, dim1):
```
Maybe after you land this change, I can make Partial Tensor a subclass of tensor too.

fduwjj commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -5,6 +5,14 @@
+    # and non-tensors, we can still return a ReplicatedTensor as the result.
+    torch.unsqueeze,
+    torch.Tensor.unsqueeze,
+    torch.Tensor.__getitem__,
```
This is for `replicated_tensor[0]`, right?

pritamdamania87 replied:
Yes more generally something like `replicated_tensor[...]`

fduwjj commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -101,7 +112,8 @@ def dispatch_arg(arg):
             if func in get_default_nowrap_functions():
                 return rs
-            if all_replicated and isinstance(rs, torch.Tensor) and not isinstance(rs, cls):
+            if (all_replicated and isinstance(rs, torch.Tensor) and not isinstance(rs, cls)) or \
```
N00b question, is the 'cls' here just `ReplicatedTensor`?

pritamdamania87 replied:
Yes, let me update it so it is more clear.

fduwjj reviewed (approved):
Overall, looks great and you might need to rebase the PR a little bit. Thanks!

wanchaol commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -5,6 +5,14 @@
 from torch.distributed import distributed_c10d
 from torch.overrides import get_default_nowrap_functions
 
+_REPLICATED_WITH_NON_TENSOR_ALLOWLIST = [
```
I think this list is for the case where all args are either ReplicatedTensor or scalars, instead of having two flags `all_replicated` and `replicated_with_non_tensor`, could we have a single flag to check if there's no `torch.Tensor` in the args (ShardedTensor or PartialTensor handles in their own __torch_function__), then we convert to ReplicatedTensor, otherwise leave it as tensor?

pritamdamania87 replied:
Hmm but couldn't there be a case where there is an op with ReplicatedTensor and PartialTensor and it gets dispatched to ReplicatedTensor's `torch_function` instead? We re-dispatch to ShardedTensor in the case of ShardedTensor, but I don't see anything similar for ReplicatedTensor.

pritamdamania87 replied:
So another issue is Replicated (op) scalar is not always guaranteed to be ReplicatedTensor. Imagine torch.add(ReplicatedTensor, x) where `x` is different on each rank.

So it seems like we do need this additional flag in which case we have look through a whitelist like this.

wanchaol reviewed (approved):
looks good to me, one suggestion to simplify the return type conversion for ReplicatedTensor

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/partial_tensor.py:
```diff
@@ -1,13 +1,45 @@
+        return wrapper
+    return decorator_sharded_func
 
 class _PartialTensor(object):
```
Yup will rebase this.

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:
```diff
@@ -0,0 +1,22 @@
+    dim = kwargs['dim']
+    sharding_dim = input.sharding_spec().dim
+    ndims = len(input.size())
+    if dim == sharding_dim or dim + ndims == sharding_dim or sharding_dim + ndims == dim:
+        exp = torch.exp(input.local_tensor())
+        exp_sum = exp.sum(dim=dim).unsqueeze(dim=dim)
+        exp_sum = torch.distributed.nn.functional.all_reduce(exp_sum, group=pg)
+        smax = torch.div(exp, exp_sum)
```
Yup will do as I rebase.

pritamdamania87 committed:
Update on "Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor."


Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please

pritamdamania87 committed:
Update on "Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor."


Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Additional ops for ShardedTensor, ReplicatedTensor and PartialTensor."


Adding the following ops:

1) softmax for ShardedTensor
2) getitem and unsqueeze for ReplicatedTensor
3) transpose and cat for PartialTensor

Differential Revision: [D35979510](https://our.internmc.facebook.com/intern/diff/D35979510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35979510/)!

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please

pritamdamania87 commented:
@pytorchbot merge this please
